English revision work by Hengrui (#362)
Oneflow-Inc/oneflow-documentation: en/docs/basics/05_autograd.md

# AUTOGRAD

> The training process of a neural network is powered by **backpropagation algorithm**. In the backpropagation process, we update the parameters by obtaining the gradient of the loss function to the parameters.

Reviewer: ```suggestion
The training process of a neural network is powered by **backpropagation algorithm**. In the backpropagation process, we update the parameters by obtaining the gradient of the loss function with respect to the parameters.
```
我看的 pytorch 的😂

> OneFlow provides an automatic differentiation engine, which can calculate the gradient of the parameters in the neural network automatically.

Reviewer: ```suggestion
OneFlow provides an autograd engine, which can calculate the gradient of the parameters in the neural network automatically.
```
其实用 automatic differentiation 更学术和准确。autograd 是一个自我发明（起源自 pytorch）的模块的名字。
不过介于这篇名字就是 Autograd，我们在这用 autograd 强调，我觉得是可以的。

> We will first introduce the basic concepts of the computational graph, which are conducive to understand the common settings and limitations of Oneflow's automatic differentiation. Then we introduce OneFlow's common automatic differentiation interfaces.【待修改】

Reviewer: 不好翻译的话，意译也可以。就是我们要先介绍计算图的概念，这有利于理解 OneFlow 的 autograd。

Reviewer: 计算图 computation graph 就可以了

Author: 已解决

##  Computational Graph

Computational graphs are composed of tensors and operators. We show this in code as below:

```python
import oneflow as flow

def loss(y_pred, y):
    return flow.sum(1/2*(y_pred-y)**2)

x = flow.ones(1, 5)  # input【不确定】
w = flow.randn(5, 3, requires_grad=True)
b = flow.randn(1, 3, requires_grad=True)
z = flow.matmul(x, w) + b

y = flow.zeros(1, 3)  # label
l = loss(z,y)
```

> Corresponding calculation diagram：

Reviewer: computation graph

Author: 已解决

![todo](./imgs/compute_graph.png)

> In computational graph, leaf nodes are the input tensors, like `x`, `w`, `b`, and `y`, and root nodes are the output tensors, like `loss`.

Reviewer: 只有输出，没有输入的节点称为 叶子节点
 the nodes only with output and with no input called **leaf node**, like ....

root 的翻译同理，目前的翻译，意义上是有偏差的。

另外这里是介绍“定义”，如果不好突出概念，修改样式也可以的，比如像上面我例子那样加粗体

Author: 已解决

During the backpropagation process, the gradient of `l` to `w` and `b` is required to update `w` and `b`. Therefore, we need to set `requires_grad` as `True` when creating them.

## Automatic Gradient

> ### `backward()` and Gradient【不确定】

Reviewer: 我觉得这样可以

> During the backpropagation process, we need to get the gradients of `l` to `w`、`b` respectively, shown as $\frac{\partial l}{\partial w}$ and $\frac{\partial l}{\partial b}$. We only need to call the 'backward()' method on `l`, and then OneFlow will automatically calculate the gradients and store them in the `w.grad` and `b.grad`.【不确定】

Reviewer: 英文没有顿号，需要把顿号改成英文里的符号（逗号）

Author: 已解决

> During the backpropagation process, we need to get the gradients of `l` to `w`、`b` respectively, shown as $\frac{\partial l}{\partial w}$ and $\frac{\partial l}{\partial b}$. We only need to call the 'backward()' method on `l`, and then OneFlow will automatically calculate the gradients and store them in the `w.grad` and `b.grad`.【不确定】

Reviewer: ```suggestion
During the backpropagation process, we need to get the gradients of `l` to `w`, `b` respectively, shown as $\frac{\partial l}{\partial w}$ and $\frac{\partial l}{\partial b}$. We only need to call the 'backward()' method of `l`, and then OneFlow will automatically calculate the gradients and store them in the `w.grad` and `b.grad`.
```

```python
l.backward()
print(w.grad)
print(b.grad)
```

```text
tensor([[0.9397, 2.5428, 2.5377],
        [0.9397, 2.5428, 2.5377],
        [0.9397, 2.5428, 2.5377],
        [0.9397, 2.5428, 2.5377],
        [0.9397, 2.5428, 2.5377]], dtype=oneflow.float32)
tensor([[0.9397, 2.5428, 2.5377]], dtype=oneflow.float32)
```

### Gradient for Non-leaf Nodes
By default, only gradients of leaf nodes with `requires_grad=True` will be retained. The 'grad' of a non-leaf node is automatically freed during the calling of 'backward' and cannot be viewed.

`Tensor.retain_grad()` can be called to retain and view the 'grad' of a non-leaf node.

```python
from math import pi
n1 = flow.tensor(pi/2, requires_grad=True)
n2 = flow.sin(n1)
n2.retain_grad()
n3 = flow.pow(n2, 2)

n3.backward()
print(n1.grad)
print(n2.grad)
```

> we get $\frac{\partial n_3}{\partial n_1}$ and $\frac{\partial n_3}{\partial n_2}$ using the code above.

Reviewer: 这些数学公式，从源码review难度太大了，你记得自己也先 mkdcos build，查看下显示的效果

Author: 这里不知道为什么一直显示的是代码，待解决

Reviewer: 之前 base 分支的缺少了必要的配置。现在已经添加，并且已经同步到你这个分支了。你在自己电脑上同步下：

```shell
git pull origin revision_en_version
```
再 `medics build` 应该就可以了

Output:

```
tensor(-8.7423e-08, dtype=oneflow.float32)
tensor(2., dtype=oneflow.float32)
```

> ### Call `backward()` Multiple Times on a Calculation Graph

Reviewer: Computation Graph

Author: 已解决

> By default, we can only call `backward()` once for each calculation graph. For example, the following code will report an error:

Reviewer: 计算图的翻译记得都自己检查下。我就不刻意重复强调了

Author: 已全部解决

> By default, we can only call `backward()` once for each calculation graph. For example, the following code will report an error:

Reviewer: ```suggestion
By default, we can only call `backward()` once for each calculation graph. For example, the following code will raise an error:
```
好像编程相关很少真实用 "report an error"

```python
n1 = flow.tensor(10., requires_grad=True)
n2 = flow.pow(n1, 2)
n2.backward()
n2.backward()
```

Error message:

> Maybe you try to backward through the node a second time. Specify retain_graph=True when calling .backward() or autograd.grad() the first time.

> If we need `backward()` multiple times on the same calculation graph, `retain_graph` needs to be set as `True` when calling.

Reviewer: ```suggestion
If we need to call `backward()` multiple times on the same computation graph, `retain_graph` needs to be `True`.
```

Author: 已解决

```python
n1 = flow.tensor(10., requires_grad=True)
n2 = flow.pow(n1, 2)

n2.backward(retain_graph=True)
print(n1.grad)
n2.backward()
print(n1.grad)
```

Output：

```text
tensor(20., dtype=oneflow.float32)
tensor(40., dtype=oneflow.float32)
```

The above output shows that OneFlow will **accumulate** the gradient calculated by `backward()` multiple times.

> We need to call the `zeros_()` to clear the gradient:

Reviewer: The `zeros_()` can be used to clear the gradient

现在的翻译，然我感觉容易误解成“必需调用 Zeros，清空梯度”

Author: 已改为By calling the `zeros_()`, we can clear the gradient:

```python
n1 = flow.tensor(10., requires_grad=True)
n2 = flow.pow(n1, 2)

n2.backward(retain_graph=True)
print(n1.grad)
n1.grad.zeros_()
n2.backward()
print(n1.grad)
```

Output：

```text
tensor(20., dtype=oneflow.float32)
tensor(20., dtype=oneflow.float32)
```

### Disabled Gradient Calculation

By default, OneFlow will trace and calculate gradients of Tensors with `requires_grad = Ture`.

> However, in some cases, we don't need OneFlow to keep tracing gradients such as just wanting to try forward prop. Then we can use [oneflow.no_grad()](https://oneflow.readthedocs.io/en/master/oneflow.html#oneflow.no_grad) or [oneflow.Tensor.detach()](https://oneflow.readthedocs.io/en/master/tensor.html#oneflow.Tensor.detach) to set.

Reviewer: ```suggestion
However, in some cases, we don't need OneFlow to keep tracing gradients such as just wanting the forward pass for inference. Then we can use [oneflow.no_grad()](https://oneflow.readthedocs.io/en/master/oneflow.html#oneflow.no_grad) or [oneflow.Tensor.detach()](https://oneflow.readthedocs.io/en/master/tensor.html#oneflow.Tensor.detach) to set.
```

```python
z = flow.matmul(x, w)+b
print(z.requires_grad)

with flow.no_grad():
    z = flow.matmul(x, w)+b
print(z.requires_grad)
```

Output：

```text
True
False
```

```python
z_det = z.detach()
print(z_det.requires_grad)
```

Output：

```text
False
```

### Gradients for Non-Scalar Outputs
Usually, we call `backward()` on scalar `loss`.

> However, if `loss` is a tensor, an error will be reported when calling `backward()` on `loss`.

Reviewer: ```suggestion
However, if `loss` is a tensor, an error will be raised when calling `backward()` on `loss`.
```

```python
x = flow.randn(1, 2, requires_grad=True)
y = 3*x + 1
y.backward()
```

Error message：

> Check failed: IsScalarTensor(*outputs.at(i)) Grad can be implicitly created only for scalar outputs

We can get the gradient after `y.sum()`.

```python
x = flow.randn(1, 2, requires_grad=True)
y = 3*x + 1
y = y.sum()
y.backward()
print(x.grad)
```

Output：

```text
tensor([[3., 3.]], dtype=oneflow.float32)
```

> Please refer to the "Extended Reading" section below for the analysis of the cause and solution of the error.

Reviewer: 是不是翻译成 Further Reading 更好？下同

Author: 已改

> Please refer to the "Extended Reading" section below for the analysis of the cause and solution of the error.

Reviewer: ```suggestion
Please refer to the "Further Reading" section below for the analysis of the cause and its solution.
```
